driver/src/driver/stablex_driver.rs

```diff
@@ -25,52 +41,77 @@ impl<'a> StableXDriver<'a> {
         orderbook_reader: &'a dyn StableXOrderBookReading,
         solution_submitter: &'a dyn StableXSolutionSubmitting,
         metrics: StableXMetrics,
+        batch_wait_time: Duration,
+        max_batch_elapsed_time: Duration,
     ) -> Self {
         StableXDriver {
             past_auctions: HashSet::new(),
             price_finder,
             orderbook_reader,
             solution_submitter,
             metrics,
+            batch_wait_time,
+            max_batch_elapsed_time,
         }
     }
 
-    #[cfg(test)]
-    fn with_past_auction(
-        price_finder: &'a mut dyn PriceFinding,
-        orderbook_reader: &'a dyn StableXOrderBookReading,
-        solution_submitter: &'a dyn StableXSolutionSubmitting,
-        metrics: StableXMetrics,
-    ) -> Self {
-        let mut driver =
-            StableXDriver::new(price_finder, orderbook_reader, solution_submitter, metrics);
-        driver.past_auctions.insert(0.into());
-        driver
+    pub fn run(&mut self) -> Result<bool> {
+        self.run_internal(SystemTime::now())
     }
 
-    pub fn run(&mut self) -> Result<bool> {
-        // Try to process previous batch auction
-        let batch_to_solve_result = self.orderbook_reader.get_auction_index();
-
-        if batch_to_solve_result
-            .as_ref()
-            .map(|batch| self.past_auctions.contains(batch))
-            .unwrap_or(false)
-        {
-            return Ok(false);
+    /// Wait until the batch id matches the batch id according to the order book
+    /// but at most until deadline.
+    fn wait_before_solving_batch(&self, batch: BatchId, deadline: Instant) {
+        let sleep_interval = Duration::from_secs(5);
+        loop {
+            if let Ok(index) = self.orderbook_reader.get_auction_index() {
+                if index.low_u64() == batch.0 {
+                    break;
+                }
+            }
+            let remaining = deadline - Instant::now();
+            if remaining <= sleep_interval {
+                thread::sleep(remaining);
+                break;
+            }
+            thread::sleep(sleep_interval);
         }
+    }
 
-        self.metrics
-            .auction_processing_started(&batch_to_solve_result);
-        let batch_to_solve = batch_to_solve_result?;
+    fn run_internal(&mut self, now: SystemTime) -> Result<bool> {
+        let open_batch = BatchId::current(now)?;
```

Comment:
I like the BatchId abstraction. Maybe we can implement previous and into<U256> directly on it to avoid accessing the internals here.